timemachine/cpp/src/gpu_utils.cu

```diff
@@ -10,3 +10,24 @@ curandStatus_t
 templateCurandNormal(curandGenerator_t generator, double *outputPtr, size_t n, double mean, double stddev) {
     return curandGenerateNormalDouble(generator, outputPtr, n, mean, stddev);
 }
+
+void __global__ k_accumulate_energy(
+    int N,
+    const __int128 *__restrict__ input_buffer, // [N]
+    __int128 *__restrict u_buffer              // [1]
+) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    // Should only have 1 idx, could have a warp that does a parallel reduce in the future
+    if (idx >= 1) {
```

Comment:
doesn't `cub::reduce` support `__int128`?

Reply:
It does, I had avoided cub initially as had imagined the cost of the setup of CUB would be larger than the potential speed up. While I was wrong about that, the parallel reduction that was added gets us within 1% of the previous benchmark. I think cub is still overkill here as it does inter-block reductions while we should be able to get away with block reductions.

The changes that I would like to follow up on after this PR:

- Make the amount of memory allocated for energy buffers constant relative to the system size. This will amortize more of the accumulation in the kernels that generate the energies as well as reduce memory
- Improve the barostat code, its janky as I didn't know what I was doing and doing more work than I think it needs to.

Reply:
sgtm!